python/cuml/test/test_random_forest.py

```diff
@@ -743,6 +744,88 @@ def test_rf_printing(capfd, n_estimators, detailed_printing):
+        cuml_model = curfc(max_features=1.0, rows_sample=1.0,
+                           n_bins=16, split_algo=0, split_criterion=0,
+                           min_rows_per_node=2, seed=23707, n_streams=1,
+                           n_estimators=n_estimators, max_leaves=-1,
+                           max_depth=max_depth)
+        y = y.astype(np.int32)
+    elif estimator_type == 'regression':
+        cuml_model = curfr(max_features=1.0, rows_sample=1.0,
+                           n_bins=16, split_algo=0,
+                           min_rows_per_node=2, seed=23707, n_streams=1,
+                           n_estimators=n_estimators, max_leaves=-1,
+                           max_depth=max_depth)
+        y = y.astype(np.float32)
+    else:
+        assert False
+
+    # Train model on the data
+    cuml_model.fit(X, y)
+
+    json_out = cuml_model.dump_as_json()
+    json_obj = json.loads(json_out)
+
+    # Test 1: Output is non-zero
+    assert '' != json_out
+
+    # Test 2: JSON object contains correct number of trees
+    assert isinstance(json_obj, list)
+    assert len(json_obj) == n_estimators
+
+    # Test 3: Traverse JSON trees and get the same predictions as cuML RF
+    def predict_with_json_tree(tree, x):
+        if 'children' not in tree:
+            assert 'leaf_value' in tree
+            return tree['leaf_value']
+        assert 'split_feature' in tree
+        assert 'split_threshold' in tree
+        assert 'yes' in tree
+        assert 'no' in tree
+        if x[tree['split_feature']] <= tree['split_threshold']:
+            return predict_with_json_tree(tree['children'][0], x)
+        return predict_with_json_tree(tree['children'][1], x)
+
+    def predict_with_json_rf_classifier(rf, x):
+        # Returns the class with the highest vote. If there is a tie, return
+        # the list of all classes with the highest vote.
+        vote = []
+        for tree in rf:
+            vote.append(predict_with_json_tree(tree, x))
+        vote = np.bincount(vote)
+        max_vote = np.max(vote)
+        majority_vote = np.nonzero(np.equal(vote, max_vote))[0]
+        return majority_vote
+
+    def predict_with_json_rf_regressor(rf, x):
+        pred = 0.
+        for tree in rf:
+            pred += predict_with_json_tree(tree, x)
+        return pred / len(rf)
+
+    if estimator_type == 'classification':
+        expected_pred = cuml_model.predict(X).astype(np.int32)
+        for idx, row in enumerate(X):
+            majority_vote = predict_with_json_rf_classifier(json_obj, row)
+            assert expected_pred[idx] in majority_vote
+    elif estimator_type == 'regression':
+        expected_pred = cuml_model.predict(X).astype(np.float32)
+        pred = []
+        for idx, row in enumerate(X):
+            pred.append(predict_with_json_rf_regressor(json_obj, row))
+        pred = np.array(pred, dtype=np.float32)
+        np.testing.assert_almost_equal(pred, expected_pred, decimal=6)
```

Comment:
Should we test that node splits are exactly identical between the binary/in-memory version and the json version? Or do we want to not care that much

Reply:
> Should we test that node splits are exactly identical between the binary/in-memory version and the json version?

This is already covered by testing equality of predictions. Trees are notoriously sensitive to small perturbations in split thresholds.